Support dynamic registration of mojo services on Android

mpcomplete opened the pull request:
SkyApplication now reads service info from a services.json file bundled
with the apk. For each service, it registers a method that invokes
connectToService on a named class. This way, third party services can
register themselves.

There's a corresponding change to flutter_tools to generate this
services.json when building an apk that depends on services.

Hixie commented on sky/shell/BUILD.gn:
```diff
@@ -193,6 +193,14 @@ if (is_android) {
     ]
   }
 
+  # TODO(mpcomplete): cargo-culting. See questions in flutter_tools's apk.dart.
```
I don't see any questions in apk.dart. Can you elaborate?

Hixie replied:
Oh, I see. This refers to the other CL.

abarth commented:
Can we convert some of the currently built-in services to this model?  For example, can we load the sensors service this way?  What about the network service?

abarth commented:
Can we load up the Java code lazily, for example when the service is first requested?  Does the services.cfg file contain enough information to let us do that in the future?  What format is the services.cfg in?  Is that format one that we can extend in the future as we decide we want to add functionality to this system?

krisgiesing commented on sky/shell/platform/android/org/domokit/sky/shell/SkyApplication.java:
```diff
@@ -63,10 +69,29 @@ protected void onBeforeResourceExtraction(ResourceExtractor extractor) {
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(manager.open(SERVICES)))) {
+            String line = null;
+            while ((line = reader.readLine()) != null)
+                registerService(registry, line);
```
Might want to filter out blank lines here, and possibly also comments.  If the file gets more complex the parsing might want to be pulled out into a separate method.

mpcomplete commented:
This is ready for another look.

To answer your questions:
> Can we convert some of the currently built-in services to this model? For example, can we load the sensors service this way? What about the network service?

Yeah, I think we could. I didn't know how to do that until I did it for GCM, so I would prefer to just keep this CL as-is. We can convert the others now that we have the technology.

> Can we load up the Java code lazily, for example when the service is first requested? Does the services.cfg file contain enough information to let us do that in the future? What format is the services.cfg in? Is that format one that we can extend in the future as we decide we want to add functionality to this system?

Good idea. It was a dumb format, so I changed it to JSON and added some more info. Now it can (a) load the Java code lazily, and (b) be extended in the future.

abarth commented on sky/shell/BUILD.gn:
```diff
@@ -220,6 +220,35 @@ if (is_android) {
+      ":shell_apk",
+      ":shell_lib",
+    ]
+  }
```
We should restructure the build system to better match what we're doing.  Specifically, we should create the dylib for the C++ and the jar for the Java and then have a step that combines them into the APK.

It's not clear whether we should distribute a jar or a dex.jar.  What's the tradeoff between the two?

mpcomplete replied:
Good suggestion. I changed the android_apk command from our rules.gni, and now it builds a classes.dex.jar before linking it into classes.dex. The .dex.jar is used by both our build system and by 'flutter apk' when linking with other services. Seems much cleaner now.

jar vs dex.jar - jar contains the java bytecode (bunch of .class files), while dex.jar has the single dalvik binary. The main tradeoff that I can see is that dex.jars seem to be smaller. But there's probably others that I don't know about.

abarth commented:
LGTM, although its making the build system slightly messy.

mpcomplete commented:
I made some changes to rules.gni and release_engine.py that make things a bit cleaner.

abarth commented:
LGTM
